### ISSUE TITLE
xkb::select_events is missing a parameter

### ISSUE THREAD

**Issue by robert-hrusecky** [psychon/x11rb#583]
The generated function looks like this:
```rust
pub fn select_events<'c, 'input, Conn, A, B, C, D>(conn: &'c Conn, device_spec: DeviceSpec, clear: A, select_all: B, affect_map: C, map: D, details: &'input SelectEventsAux) -> Result<VoidCookie<'c, Conn>, ConnectionError>
where
    Conn: RequestConnection + ?Sized,
    A: Into<u16>,
    B: Into<u16>,
    C: Into<u16>,
    D: Into<u16>,

```
But the corresponding XCB function looks like this:
```C
xcb_void_cookie_t
xcb_xkb_select_events (xcb_connection_t      *c,
                       xcb_xkb_device_spec_t  deviceSpec,
                       uint16_t               affectWhich,
                       uint16_t               clear,
                       uint16_t               selectAll,
                       uint16_t               affectMap,
                       uint16_t               map,
                       const void            *details);

```

The `affect_which` parameter is missing. It seems like a similar issue (#281) was fixed already for the old python generator.

**Comment by psychon** [comment]
TL;DR: You decide on the value of the `affectWhich` argument via `SelectEventsAux`. In XCB, there is a `void *` last argument from which XCB just copies as many bytes as it wants. We wanted a memory-safe API instead.

Long version:

You already found #281. The other important part is #450. This PR explicitly removed the `affect_which` argument that you are looking for. Let's look at some XML...

The "magic" which detects length fields triggers here. For example, here is how `InternAtom` is defined in the XML:
```xml
  <request name="InternAtom" opcode="16">
    <field type="BOOL" name="only_if_exists" />
    <field type="CARD16" name="name_len" />
    <pad bytes="2" />
    <list type="char" name="name">
      <fieldref>name_len</fieldref>
    </list>
    <reply>
[...]
    </reply>
    <doc>
[...]
    </doc>
  </request>
```
Clearly, `name_len` is just the length of the `name` argument. Since Rust slices already come with a length, this parameter is passed implicitly. Thus, the Rust API is: `pub fn intern_atom<'c, 'input, Conn>(conn: &'c Conn, only_if_exists: bool, name: &'input [u8]) -> Result<Cookie<'c, Conn, InternAtomReply>, ConnectionError>`. No `name_len` argument (once upon a time, we did have a `name_len` argument here...)

Okay, the above is an obvious case. Here comes a less obvious one:
```xml
	<request name="SelectEvents" opcode="1">
		<field name="deviceSpec" type="DeviceSpec" />
		<field name="affectWhich" type="CARD16" mask="EventType" />
		<field name="clear" type="CARD16" mask="EventType" />
		<field name="selectAll" type="CARD16" mask="EventType" />
		<field name="affectMap" type="CARD16" mask="MapPart" />
		<field name="map" type="CARD16" mask="MapPart" />
		<switch name="details">
			<op op="&amp;">
				<fieldref>affectWhich</fieldref>
				<op op="&amp;">
					<unop op="~"><fieldref>clear</fieldref></unop>
					<unop op="~"><fieldref>selectAll</fieldref></unop>
				</op>
			</op>
[omitted all the bitcases]
		</switch>
	</request>
```
The above says that there are multiple cases (`<switch>`), depending on which bits are set on the expression `affectWhich & !clear & !selectAll` (`<op op="&amp;">` is "bitwise and" and `<unop op="~">` is "bitwise not").
The value for the `<switch>` is passed in via the `details` argument in the API: You already select which of the bit cases are present. Thus, the API caller already provides the value of the expression `affectWhich & !clear & !selectAll`. This makes the "length field deduction" trigger and remove an unnecessary / implicit argument from the function. (Okay, "length field deduction" is a bad name for this since it is not about the length of something, but hopefully you still get the idea. I do not know how the code generator currently calls this.)

Anyway, now that we already have the value of `affectWhich & !clear & !selectAll`, we have a problem: If all of these three fields are also present as API arguments, the user could try to send invalid requests where the value provided for the `<switch>` does not match the values provided for these three fields. Thus, one of these fields was chosen to be implicit. Its value is deduced from the value for the `<switch>`. This argument is the `affect_which` field.

Here is how the generated code computes this field:
```rust
        let affect_which = u16::try_from(self.details.switch_expr() | (u32::from(self.clear) | u32::from(self.select_all))).unwrap();
```
So... we have the value `affectWhich & !clear & !selectAll` let's call this `switch_expr`. We know from something that is not part of the XML (see #450) that `(clear | selectAll) & !affectWhich == 0` must hold, i.e. all bits that are set in either `clear` or `selectAll` must also be set in `affectWhich`. This last property means that `affectWhich = switch_expr | clear | select_all`, which is exactly what the code generator generates. Hence, we can deduce the value of `affectWhich` from the other arguments.

Hopefully this makes it clearer what is going on.

Sadly, the XML does not provide names for these bit cases, so we end up enumerating them, leading to this ugly struct that you now have to work with (`Option` is used to decide between "this case is present" / "this case is not present") (also see #156):
```
pub struct SelectEventsAux {
    pub bitcase1: Option<SelectEventsAuxBitcase1>,
    pub bitcase2: Option<SelectEventsAuxBitcase2>,
    pub bitcase3: Option<SelectEventsAuxBitcase3>,
    pub bitcase4: Option<SelectEventsAuxBitcase4>,
    pub bitcase5: Option<SelectEventsAuxBitcase5>,
    pub bitcase6: Option<SelectEventsAuxBitcase6>,
    pub bitcase7: Option<SelectEventsAuxBitcase7>,
    pub bitcase8: Option<SelectEventsAuxBitcase8>,
    pub bitcase9: Option<SelectEventsAuxBitcase9>,
    pub bitcase10: Option<SelectEventsAuxBitcase10>,
    pub bitcase11: Option<SelectEventsAuxBitcase11>,
}
```
If you are looking at some C code using XCB's XKB API, I could help translating that to x11rb.

**Comment by robert-hrusecky** [comment]
Yes, that clears it up, thanks.



> If you are looking at some C code using XCB's XKB API, I could help translating that to x11rb.

Thanks, but I was just looking at some C code examples to try and figure out how the request is supposed to work since I found XKB spec hard to understand and the XCB man pages are unhelpful.
